aten/src/ATen/native/mkldnn/xpu/detail/WoQMatmul.cpp

```diff
@@ -0,0 +1,178 @@
+#include <c10/xpu/XPUFunctions.h>
+
+#include <ATen/native/mkldnn/xpu/detail/Attr.h>
+#include <ATen/native/mkldnn/xpu/detail/Utils.h>
+
+#include <oneapi/dnnl/dnnl.hpp>
+#include <cstdint>
+
+namespace at::native::onednn {
+
+sycl::event woq_matmul_int4(
+    Tensor& result, // torchao: [M, K], dtype: fp16
+    const Tensor& mat1_, // torchao: [M, K], dtype: fp16
+    const Tensor& mat2_, // torchao quantized weight, [K/8, N], dtype: uint4x8
+    const Tensor& scale, // torchao: [K/group_size, N], dtype: fp16
+    const Tensor& zp, // torchao: [K/group_size, N], dtype: int8
+    int64_t group_size,
+    Attr attr,
+    const std::vector<sycl::event>& deps) {
+  size_t dims = result.dim();
+  TORCH_CHECK(
+      dims == 2, "INT4 matmul at XPU only works with 2D input, got ", dims);
+  TORCH_CHECK(result.defined(), "oneDNN matmul result should be defined");
+
+  at::Device cur_device = at::Device(at::kXPU, at::xpu::current_device());
+  auto engine = GpuEngineManager::Instance().get_engine(cur_device);
+  auto stream = GpuStreamManager::Instance().get_stream();
```

Comment:
A potential bug is here: the current device may not be the same as the input parameter. Add `torch_check` to check the device sameness of the inputs and determine if the input device is the same as the current device.

Reply:
Added now, thanks for suggestion.

Reply:
hi @EikanWang  https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/LinearAlgebra.cpp#L3461 I added check here